crates/provider-keyvalue-nats/src/lib.rs

```diff
@@ -0,0 +1,682 @@
+        &self,
+        cfg: NatsConnectionConfig,
+    ) -> anyhow::Result<NatsClientBundle> {
+        let mut opts = match (cfg.auth_jwt, cfg.auth_seed) {
+            (Some(jwt), Some(seed)) => {
+                let seed = KeyPair::from_seed(&seed).context("failed to parse seed key pair")?;
+                let seed = Arc::new(seed);
+                async_nats::ConnectOptions::with_jwt(jwt, move |nonce| {
+                    let seed = seed.clone();
+                    async move { seed.sign(&nonce).map_err(async_nats::AuthError::new) }
+                })
+            }
+            (None, None) => async_nats::ConnectOptions::default(),
+            _ => bail!("must provide both jwt and seed for jwt authentication"),
+        };
+        if let Some(tls_ca) = &cfg.tls_ca {
+            opts = add_tls_ca(tls_ca, opts)?;
+        } else if let Some(tls_ca_file) = &cfg.tls_ca_file {
+            let ca = fs::read_to_string(tls_ca_file)
+                .await
+                .context("failed to read TLS CA file")?;
+            opts = add_tls_ca(&ca, opts)?;
+        }
+
+        // Get the cluster_uri
+        let uri = cfg.cluster_uri.unwrap_or_default();
+
+        // Connect to the NATS server
+        let client = opts
+            .name("NATS Key-Value Provider") // allow this to show up uniquely in a NATS connection list
+            .connect(uri.clone())
+            .await?;
+
+        // Connect to JetStream
+        let js_context = if let Some(domain) = cfg.js_domain.as_ref() {
+            async_nats::jetstream::with_domain(client.clone(), domain)
+        } else {
+            async_nats::jetstream::new(client.clone())
+        };
+
+        // Set the atomic lock prefix
+        let atomic_lock_suffix = format!("{}@{}", cfg.js_domain.unwrap_or_default(), uri);
+
+        Ok(NatsClientBundle { client, js_context, atomic_lock_suffix })
+    }
+
+    /// Helper function to get the NATS client bundle from the client component's context
+    async fn get_nats_client_bundle(&self, context: Option<Context>) -> anyhow::Result<NatsClientBundle> {
+        if let Some(ref source_id) = context
+            .as_ref()
+            .and_then(|Context { component, .. }| component.clone())
+        {
+            let components = self.consumer_components.read().await;
+            let nats_bundle = match components.get(source_id) {
+                Some(nats_bundle) => nats_bundle,
+                None => {
+                    error!("consumer component not linked: {source_id}");
+                    bail!("consumer component not linked: {source_id}")
+                }
+            };
+            Ok((*nats_bundle).clone())
+        } else {
+            error!("no consumer component in request");
+            bail!("no consumer component in request")
+        }
+    }
+
+    /// Helper function to open an existing NATS Kv bucket from the client component's context
+    #[instrument(level = "debug", skip_all)]
+    async fn kv_store(
```

Comment:
I spent a lot of time trying to determine if I'm on the right track with this very same topic. As you know very well, the store (NATS bucket) is tied to the `js_domain`. In addition, a user can call the `open` function between various invocations to work with different stores available through that js_domain. So, unless, the provider can somehow determine the handle from the passed bucket name (which perhaps also mean something else is taking care of opening the store), I don't know how this could be implemented differently.

In regards to storing it in the hashmap, I considered that; it could be stored in the hashmap, but still every call requires a check to validate if what's been stored is the latest _store name_ passed, or we treat each call as independent, as I've done here; I know they have different costs and aren't the same; so, if I get guidance about where the `open` operation will, or supposed to, be handled, then this change will be easy to make.